### PR TITLE
Fix issues where internal pattern matching state was being copied

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.TextChunk.cs [hunk 1]
```diff
@@ -32,7 +32,10 @@ private struct TextChunk : IDisposable
             /// </summary>
             public TemporaryArray<TextSpan> PatternHumps;
 
-            public readonly WordSimilarityChecker? SimilarityChecker;
+            /// <summary>
+            /// Not readonly as this value caches data within it, and so it needs to be able to mutate.
+            /// </summary>
+            public WordSimilarityChecker SimilarityChecker;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.TextChunk.cs:38]: the alternative is to accept that thsi caching optimization was never working, remove it entirely, and make all this stuff readonly.

**Comment by CyrusNajmabadi** [comment]
@Cosifne ptal :)

**Comment by CyrusNajmabadi** [comment]
@genlu @Cosifne @akhera99 ptal :)

**Comment by CyrusNajmabadi** [comment]
@genlu @Cosifne @akhera99 ptal :)

**Comment by CyrusNajmabadi** [comment]
@genlu @Cosifne @akhera99 ptal :)

**Review by akhera99 — APPROVED** [review]

**Pull request merged**